DOCS-1198: Improve examples and advanced MR (#2079)
viamrobotics/docs: docs/modular-resources/advanced/create-subtype.md

weight: 30
type: "docs"
tags: ["rdk", "extending viam", "modular resources", "API"]
description: "Define a new API for a resource that does not fit into existing component or service subtypes."
no_list: true
aliases:
  - "/extend/modular-resources/create/create-subtype/"
---

You can define a new resource _{{< glossary_tooltip term_id="subtype" text="subtype" >}}_ API if:

- You have a {{% glossary_tooltip term_id="resource" text="resource" %}} that does not fit into any of the existing {{< glossary_tooltip term_id="component" text="component" >}} or {{< glossary_tooltip term_id="service" text="service" >}} subtypes.
- You have a resource that could fit into an existing subtype, but you want to define an API with different methods and messages than the ones in the existing [APIs](/program/apis/) for that subtype.

>   {{% alert title="Tip" color="tip" %}}

Reviewer: ```suggestion

  {{% alert title="Tip" color="tip" %}}
```

   If you are writing your module using Python, you can use this [module generator tool](https://github.com/viam-labs/generator-viam-module) to generate stub files for the new API as well as a new {{< glossary_tooltip term_id="module" text="module" >}} that implements the new API.

{{% /alert %}}
   {{% /alert %}}

4. Define your new API:
1. Define your new API:

   - [Write the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
   - [Create the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
     For reference:
     - [Example modular component proto file](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/complex_module/src/proto/gizmo.proto)
     - [Example modular service proto file](https://github.com/viam-labs/speech/blob/main/src/speech/proto/speech.proto)
     - [Built-in Viam resource proto files for reference](https://github.com/viamrobotics/api/tree/main/proto/viam)

>    - And define the proto methods in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go`.

Author: I think this is the same https://docs.viam.com/modular-resources/create/#code-a-new-resource-model? So I've moved it to Next steps? Is that wrong?

Reviewer 2: Seems correct to me but @JessamyT could confirm

Reviewer: No I do not think that is right. Creating a model doesn't involve proto methods does it? This is the underlying API definition stuff, not the model implementation, as far as I know. That's why we called them "api.py" or whatever--to indicate that they're part of the API definition as opposed to the model definition.

Author: ok I've put it back

> Implement the proto methods you defined in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go`.

Reviewer 2: Does it have to be called `api.py` or `api.go`?

> Implement the proto methods you defined in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go`.

Author: ```suggestion
Implement the proto methods you defined in a protobuf-supported language.
If you use Python call your file <file>apy.py</file>.
If you use Go,call your file `api.go`.
```

weight: 30
type: "docs"
tags: ["rdk", "extending viam", "modular resources", "API"]

> description: "Define a new resource subtype and API for a resource that does not fit into existing component or service subtypes."

Reviewer: Ooh that description got incorrected--good it's getting fixed. But, I think "subtype and API" is redundant; how about:
```suggestion
description: "Define a new API for a resource that does not fit into existing component or service subtypes."
```

no_list: true
aliases:
  - "/extend/modular-resources/create/create-subtype/"
---

> You can define a new resource _{{< glossary_tooltip term_id="subtype" text="subtype" >}}_ (and an API for that subtype) if:

Reviewer: ```suggestion
You can define a new resource _{{< glossary_tooltip term_id="subtype" text="subtype" >}}_ API if:
```
[opt] I realize these are not new words but maybe this would be a better way to put it that is less redundant? Because a new subtype _is_ a new API?

- You have a {{% glossary_tooltip term_id="resource" text="resource" %}} that does not fit into any of the existing {{< glossary_tooltip term_id="component" text="component" >}} or {{< glossary_tooltip term_id="service" text="service" >}} subtypes.
- You have a resource that could fit into an existing subtype, but you want to define an API with different methods and messages than the ones in the existing [APIs](/program/apis/) for that subtype.

>   For example, you have a [sensor](/components/sensor/) and you want to define a `Calibrate` method.

Reviewer: Maybe we should actually remove this--it might be a good example for what you should do with DoCommand. Dunno what happened with the de-emphasis ticket but this might tie into that.

  {{% alert title="Tip" color="tip" %}}

  If you want to use most of an existing API but need just a few other functions, try using the [`DoCommand`](/program/apis/#docommand) endpoint and [extra parameters](/program/use-extra-params/) to add custom functionality to an existing subtype.

  {{% /alert %}}

## Define your new resource API

Viam uses [protocol buffers](https://protobuf.dev/) for API definition.

> To define a new subtype, you need to define the methods and messages of the new API in [protobuf](https://github.com/protocolbuffers/protobuf), generate all necessary [protobuf module files](https://buf.build/docs/generate/usage/), and write code in Python or Go to implement the higher level server and client functions required.

Reviewer: Why this change? I'd do it in the originally-written order I think. I *think* it'd be easier to figure out what you need by writing the actual code, and then making the protobuf to fit later. But maybe either is possible? @mcvella

The following steps guide you through this process in more detail:

1. Decide whether your subtype is a {{< glossary_tooltip term_id="component" text="component" >}} or a {{< glossary_tooltip term_id="service" text="service" >}}.
   If it provides an interface to control hardware, it is a component.
   If it provides higher-level functionality, it is a service.
1. Choose a name for your subtype.

>    For example, "gizmo".

Reviewer: If it's not monospace, wouldn't it follow normal punctuation guidelines?

Author: I am happy to change this to monospace. I do not want to risk people putting periods in the names of their components because they are confused
```suggestion
   For example, `gizmo`.
```

1. Create a directory for your module.
   Within that, create a directory called <file>src</file>.

   {{% alert title="Tip" color="tip" %}}

   If you are writing your module using Python, you can use this [module generator tool](https://github.com/viam-labs/generator-viam-module) to generate stub files for the new API as well as a new {{< glossary_tooltip term_id="module" text="module" >}} that implements the new API.

   {{% /alert %}}

1. Define your new API:

>    - [Create the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.

Reviewer: I think "write" makes it more clear that you're actually typing these out yourself, as opposed to the auto-generated ones in other steps.
```suggestion
   - [Write the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
```

{{% /alert %}}
   {{% /alert %}}

4. Define your new API:
1. Define your new API:

   - [Write the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
   - [Create the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
     For reference:
     - [Example modular component proto file](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/complex_module/src/proto/gizmo.proto)
     - [Example modular service proto file](https://github.com/viam-labs/speech/blob/main/src/speech/proto/speech.proto)
     - [Built-in Viam resource proto files for reference](https://github.com/viamrobotics/api/tree/main/proto/viam)
   - And define the proto methods in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go`.
     - [Example component in Python](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/complex_module/src/gizmo/api.py)

>      - [Example service in Python](https://github.com/viam-labs/speech/blob/main/src/speech/api.py)

Reviewer: Leave these in too

>      All files except <file>speech.proto</file> were generated.

Reviewer: I'm not sure this is true-- @mcvella ?
```suggestion
     The `buf.` files were generated.
     The <file>speech.proto</file> was manually written.
```

> - [Example service in Python](https://github.com/viam-labs/speech/blob/main/src/speech/api.py)

Reviewer: As discussed above, move these back (unless I'm quite confused)

   If it provides higher-level functionality, it is a service.
1. Choose a name for your subtype.
   For example, `gizmo`.
1. Create a directory for your module.
   Within that, create a directory called <file>src</file>.

   {{% alert title="Tip" color="tip" %}}

   If you are writing your module using Python, you can use this [module generator tool](https://github.com/viam-labs/generator-viam-module) to generate stub files for the new API as well as a new {{< glossary_tooltip term_id="module" text="module" >}} that implements the new API.

   {{% /alert %}}

1. Define your new API:

   - [Write the proto](https://protobuf.dev/programming-guides/proto3/) methods in a `<subtype name>.proto` file inside your <file>src/proto</file> directory.
     For reference:
     - [Example modular component proto file](https://github.com/viamrobotics/viam-python-sdk/blob/main/examples/complex_module/src/proto/gizmo.proto)
     - [Example modular service proto file](https://github.com/viam-labs/speech/blob/main/src/speech/proto/speech.proto)
     - [Built-in Viam resource proto files](https://github.com/viamrobotics/api/tree/main/proto/viam)

>    - And define the proto methods in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go` respectively.

Reviewer: ```suggestion
   - And define the proto methods in a protobuf-supported language such as Python or Go in a file called `api.py` or `api.go`, respectively.
```
[nit]
